### ISSUE TITLE
Training Step of GRPO in Wandb.

### ISSUE THREAD

**Issue by mandyyyyii** [huggingface/trl#4004]
I’m training with the following setup: dataset size = 10k, batch size = 32, 8 GPUs, gradient accumulation = 4, num_generation = 8, and epoch = 1.
By my calculation, the total number of optimization steps should be:
10k×8/32=2.5k global steps. In wandb, however, I see that reward and min/max length are logged over ~40k steps, while training loss aligns with my expectation at ~2.5k steps. I’m confused why these metrics are tracked as if there were 40k training steps. If logging corresponds to every model generation, shouldn’t it actually be ~80k instead?

Additionally, when I implement a custom GRPO trainer, I generate once per batch, then run another generation with a different prompt (but only logged the information of first generation --reward/length-- using the same code). In this case, wandb shows ~42.5k steps — slightly higher than 40k — as if each batch adds some extra logging offset.
Has anyone else encountered similar discrepancies?

**Comment by LeonEricsson** [comment]
> batch size = 32

 is this referring to `per_device_train_batch_size`? The effective batch size is given by  `per_device_train_batch_size * num_gpus * gradient_accumulation`, which would be `32*8*4=1024`. Meaning you should only be doing `10e3*8/1024≈79` optimization steps. So something seems off; could you supply the entire config

**Comment by mandyyyyii** [comment]
Hi, thanks for your reply. 32 refer to the effective batch size. per_device_train_batch_size=1 in my case. Below is my config: 

accelerate launch --config_file gpu4.yaml grpo_ref.py \
    --model_name_or_path "meta-llama/Llama-3.1-8B-Instruct" \
    --dataset_config "main" \
    --dataset_train_split "train" \
    --dataset_test_split "test" \
    --learning_rate 5e-7 \
    --beta 0 \
    --lr_scheduler_type cosine \
    --max_grad_norm 1.0 \
    --optim adamw_torch \
    --seed 42 \
    --weight_decay 0.01 \
    --num_train_epochs 1 \
    --num_generations 8 \
    --data_seed 42 \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --gradient_checkpointing \
    --max_completion_length 2048 \
    --vllm_mode "colocate" \
    --temperature 1 \
    --use_vllm True \
    --mask_truncated_completions True \
    --logging_steps 10 \
    --save_strategy "no" \
    --bf16 True \
    --report_to wandb \

**Comment by LeonEricsson** [comment]
unfortunately I'm having a hard time replicating your observed behaviour. I always see training loss, reward, min/max completions etc logged at the expected (same) rate of $\text{num samples} \times \text{num generations} / \text{effective batch size}$

**Comment by qgallouedec** [comment]
Hey @mandyyyyii, sorry for the very slow reply, catching up on the backlog.

Looking back at this, as @LeonEricsson said there isn't quite enough to investigate: no `trl env` output, no minimal script that reproduces the discrepancy, and the second paragraph mixes in a custom GRPO trainer where logging is your own. Closing for now, but very happy to re-open if you can share `trl env` plus a minimal repro on a current TRL version.
